File: 프로그래머스/2/12952. N－Queen/N－Queen.java
class Solution {
    public int solution(int n) {
        return queen(0, n, new int[n]);
    }

    public int queen(int s, int n, int[] board) {
        if (s == n) return 1; // 모든 퀸이 배치된 경우

        int answer = 0;
        for (int i = 0; i < n; i++) {
            board[s] = i; // 현재 행 s에 퀸을 배치
            boolean isValid = true;
            for (int j = 0; j < s; j++) {
                // 같은 열이거나 대각선이 아닌지 검사
                if (board[j] == board[s] || Math.abs(board[s] - board[j]) == s - j) {
                    isValid = false;
                    break;
                }
            }

            if (isValid) {
                answer += queen(s + 1, n, board); // 다음 행으로 재귀 호출
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def queen(s, n, l):
    # 끝줄까지 탐색했으면 return 1
    if s == n:
        return 1
    answer = 0
    # 퀸 채워넣기
    for i in range(n):
        l[s] = i
        # 가능한지 체크
        for r in range(s):
            # 같은줄 or 대각선이면 실패
            if l[r] == l[s] or abs(l[s] - l[r]) == s - r:
                break
        # 모두 통과면 다음
        else:
            answer += queen(s + 1, n, l)
    return answer

def solution(n): 
    return queen(0, n, [0] * n)